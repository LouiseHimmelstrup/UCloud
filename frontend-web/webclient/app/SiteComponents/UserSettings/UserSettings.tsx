import * as React from "react";
import { Form, Input, Button } from "semantic-ui-react";


interface UserSettingsState {
    fields: {
        oldPassword: string
        newPassword: string
        repeatedPassword: string
    }
    error: boolean
}
class UserSettings extends React.Component<{}, UserSettingsState> {
    constructor(props) {
        super(props);
        this.state = {
            fields: {
                oldPassword: "",
                newPassword: "",
                repeatedPassword: ""
            },
            error: false
        }
    }

    updateField(fieldName: string, value: string): void {
        
    }

    validateAndSubmit(): void {
        
    }    

    render() {
        return (
            <React.StrictMode>
                <Form>
                    <Form.Field label="Old password">
                        <Input placeholder="Old password" />
                    </Form.Field>
                    <Form.Field label="New password">
                        <Input placeholder="New password" />
                    </Form.Field>
                    <Form.Field label="Repeat password">
                        <Input placeholder="Repeat password" />
                    </Form.Field>
                    <Button type="button" onClick={() => this.validateAndSubmit()} content="Submit"/>
                </Form>
            </React.StrictMode>
        );
    }
}

export default UserSettings;