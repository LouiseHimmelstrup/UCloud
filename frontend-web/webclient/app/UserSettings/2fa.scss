.badge {
    height: 64px;
}