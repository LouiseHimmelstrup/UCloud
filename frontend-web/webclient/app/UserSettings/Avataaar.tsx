import { default as Avataaar } from "avataaars";
import * as React from "react";
import * as Options from "./AvatarOptions";
import { MainContainer } from "MainContainer/MainContainer";
import { Select, Label, Box, Flex } from "ui-components";


interface AvataaarModificationState {
    top: Options.TopOptions
    accessories: Options.TopAccessoryOptions
    hairColor: Options.HairColorOptions
    facialHair: Options.FacialHairOptions
    facialHairColor: Options.FacialHairColorOptions
    clothes: Options.ClothesOptions
    clothesFabric: Options.ColorFabricOptions
    clothesGraphic: Options.ClothesGraphicOptions
    eyes: Options.EyeOptions
    eyebrow: Options.EyeBrowOptions
    mouth: Options.MouthTypeOptions
    skin: Options.SkinColorOptions
}

class AvataaarModification extends React.Component<void, AvataaarModificationState> {
    constructor(props) {
        super(props);
        this.state = {
            top: Options.Top.NoHair,
            accessories: Options.TopAccessory.Blank,
            hairColor: Options.HairColor.Auburn,
            facialHair: Options.FacialHair.BeardMedium,
            facialHairColor: Options.FacialHairColor.Auburn,
            clothes: Options.Clothes.BlazerShirt,
            clothesFabric: Options.ColorFabric.Black,
            clothesGraphic: Options.ClothesGraphic.Bat,
            eyes: Options.Eyes.Cry,
            eyebrow: Options.Eyebrows.DefaultNatural,
            mouth: Options.MouthTypes.Concerned,
            skin: Options.SkinColors.Pale
        }
    }

    private extractAvataaar() {
        return {

        }
    }

    render() {
        const { ...state } = this.state;
        return (
            <MainContainer


                main={
                    <>
                        <Flex>
                            <Box ml="auto" />
                            <Avataaar
                                style={{ height: "150px" }}
                                avatarStyle="circle"
                                topType={state.top}
                                accessoriesType={state.accessories}
                                hairColor={state.hairColor}
                                facialHairType={state.facialHair}
                                facialHairColor={state.facialHairColor}
                                clotheType={state.clothes}
                                clotheColor={state.clothesFabric}
                                graphicType={state.clothesGraphic}
                                eyeType={state.eyes}
                                eyebrowType={state.eyebrow}
                                mouthType={state.mouth}
                                skinColor={state.skin}
                            />
                            <Box mr="auto" />
                        </Flex>
                        <AvatarSelect
                            update={value => this.setState(() => ({ top: value }))}
                            options={Options.Top}
                            title="Top"
                            disabled={false}
                        />
                        <AvatarSelect
                            update={value => this.setState(() => ({ accessories: value }))}
                            options={Options.TopAccessory}
                            title="Accessories"
                            disabled={state.top === "Eyepatch"}
                        />
                        <AvatarSelect
                            update={value => this.setState(() => ({ hairColor: value }))}
                            options={Options.HairColor}
                            title="Hair color"
                            disabled={!state.top.includes("Long") || state.top === "LongHairFrida"}
                        />
                        <AvatarSelect
                            update={value => this.setState(() => ({ facialHair: value }))}
                            options={Options.FacialHair}
                            title="Facial Hair"
                            disabled={state.top === "Hijab"}
                        />
                        <AvatarSelect
                            update={value => this.setState(() => ({ facialHairColor: value }))}
                            options={Options.FacialHairColor}
                            title="Facial Hair Color"
                            disabled={state.facialHair === "Blank"}
                        />
                        <AvatarSelect
                            update={value => this.setState(() => ({ clothes: value }))}
                            options={Options.Clothes}
                            title="Clothes"
                            disabled={false}
                        />
                        <AvatarSelect
                            title="Clothes Fabric"
                            options={Options.ColorFabric}
                            update={value => this.setState(() => ({ clothesFabric: value }))}
                            disabled={state.clothes === "BlazerShirt" || state.clothes === "BlazerSweater"}
                        />
                        <AvatarSelect
                            title="Graphic"
                            update={value => this.setState(() => ({ clothesGraphic: value }))}
                            options={Options.ClothesGraphic}
                            disabled={state.clothes !== "GraphicShirt"}
                        />
                        <AvatarSelect
                            title="Eyes"
                            options={Options.Eyes}
                            update={value => this.setState(() => ({ eyes: value }))}
                            disabled={false}
                        />
                        <AvatarSelect
                            title="Eyebrow"
                            options={Options.Eyebrows}
                            update={value => this.setState(() => ({ eyebrow: value }))}
                            disabled={false}
                        />
                        <AvatarSelect
                            title="Mouth type"
                            options={Options.MouthTypes}
                            update={value => this.setState(() => ({ mouth: value }))}
                            disabled={false}
                        />
                        <AvatarSelect
                            title={"Skin color"}
                            options={Options.SkinColors}
                            update={value => this.setState(() => ({ skin: value }))}
                            disabled={false}
                        />
                    </>
                }
            />)
    }
}

interface AvatarSelect<T> {
    update: (value: keyof T) => void
    options: T
    title: string
    disabled: boolean
}

function AvatarSelect<T>({ update, options, title, disabled }: AvatarSelect<T>) {
    if (disabled) return null;
    return (
        <Label mt="0.8em">{title}
            <Select onChange={({ target: { value } }) => update(value as keyof T)}>
                {Object.keys(options).map(it => <option key={it}>{it}</option>)}
            </Select>
        </Label>
    )
}

export default AvataaarModification;