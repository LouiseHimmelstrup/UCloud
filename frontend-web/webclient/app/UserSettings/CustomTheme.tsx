import {Box, Flex, Heading} from "@/ui-components";
import * as React from "react";

// TODO(Jonas): {
//      - add gradient colors;    
// }

export const CUSTOM_THEME_COLOR_KEY = "custom-theme-color-key";
export function findCustomThemeColorOnLaunch() {
    const root = document.querySelector(":root")!;
    const color = localStorage.getItem(CUSTOM_THEME_COLOR_KEY);
    if (!color) return;
    root["style"].setProperty("--blue", color);
}

type HexColor = `#${string}`;
const COLORS: HexColor[] = ["#006aff", "#bed730", "#80007d", "#eaa621"];
export function CustomTheming(): React.ReactNode {

    const root = React.useMemo(() => document.querySelector(':root')!, []);

    const setColor = React.useCallback((color: HexColor) => {
        localStorage.setItem(CUSTOM_THEME_COLOR_KEY, color);
        root["style"].setProperty('--blue', color);
    }, []);

    const activeColor = localStorage.getItem(CUSTOM_THEME_COLOR_KEY) ?? COLORS[0];

    return <div>
        <Heading>Theme color options (beta)</Heading>
        <Flex mt="12px">
            {COLORS.map(color =>
                <ColorOption key={color} isActive={color === activeColor} color={color} setColor={setColor} />
            )}
        </Flex>
    </div>
}

function ColorOption({color, setColor, isActive}: {isActive: boolean, color: HexColor, setColor(color: HexColor): void}) {
    return <Box
        onClick={() => setColor(color)}
        style={{border: "2px solid " + isActive ? "var(--black)" : "var(--white)"}}
        borderRadius="12px"
        backgroundColor={color}
        cursor="pointer"
        width="24px"
        height="24px"
        mr="12px"
    />
}
