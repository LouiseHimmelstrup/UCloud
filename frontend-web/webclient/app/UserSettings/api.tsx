import {buildQueryString} from "Utilities/URIUtilities";
import {BulkRequest} from "UCloud";

export interface UserSession {
    ipAddress: string;
    userAgent: string;
    createdAt: number;
}

export type ListUserSessionParameters = PaginationRequest;

export function listUserSessions(
    parameters: ListUserSessionParameters
): APICallParameters<ListUserSessionParameters> {
    return {
        reloadId: Math.random(),
        method: "GET",
        path: buildQueryString("/auth/sessions", parameters),
        parameters,
        context: ""
    };
}
/*
export interface EmailSettings {
    //Grant applications
    newGrantApplication: boolean;
    grantAutoApprove: boolean;
    grantApplicationUpdated: boolean;
    grantApplicationApproved: boolean;
    grantApplicationRejected: boolean;
    grantApplicationWithdrawn: boolean;
    newCommentOnApplication: boolean;
    applicationTransfer: boolean;
    applicationStatusChange: boolean;
    //Project
    projectUserInvite: boolean;
    projectUserRemoved: boolean;
    verificationReminder: boolean;
    userRoleChange: boolean;
    userLeft: boolean;
    lowFunds: boolean;
}

export interface RetrieveEmailSettingsRequest {
    username?: string
}

export function retrieveEmailSettings(
    request: RetrieveEmailSettingsRequest
):APICallParameters<RetrieveEmailSettingsRequest> {
    return {
        context: "",
        method: "GET",
        path: "/api/mail" + "/retrieveEmailSettings",
        parameters: request,
        reloadId: Math.random(),
    }
}

export interface ToggleEmailSettingsRequestItem {
    username?: string,
    settings: EmailSettings
}
*/
//export function toggleEmailSettings(
//    request: BulkRequest<ToggleEmailSettingsRequestItem>
//): APICallParameters<BulkRequest<ToggleEmailSettingsRequestItem>, any /* unknown */> {
//    return {
//        context: "",
//        method: "POST",
//        path: "/api/mail" + "/toggleEmailSettings",
//        parameters: request,
//        reloadId: Math.random(),
//        payload: request,
//    };
//}

export function invalidateAllSessions(): APICallParameters {
    return {
        reloadId: Math.random(),
        method: "DELETE",
        path: "/auth/sessions",
        context: "",
        withCredentials: true
    };
}
