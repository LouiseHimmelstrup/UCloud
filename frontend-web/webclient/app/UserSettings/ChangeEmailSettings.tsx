import {useCloudCommand} from "Authentication/DataHook";
import {Client} from "Authentication/HttpClientInstance";
import * as React from "react";
import {useCallback, useEffect, useRef} from "react";
import {Box, Button, Checkbox, Input, Label} from "ui-components";
import * as Heading from "ui-components/Heading";
import {snackbarStore} from "Snackbar/SnackbarStore";
import {EmailSettings, retrieveEmailSettings, toggleEmailSettings} from "UserSettings/api";
import {bulkRequestOf} from "DefaultObjects";
import {file} from "UCloud";
import stat = file.stat;

interface UserDetailsState {
    settings: EmailSettings
}

export const defaultEmailSettings: EmailSettings = {
    newGrantApplication: true,
    grantAutoApprove: true,
    grantApplicationUpdated: true,
    grantApplicationApproved: true,
    grantApplicationRejected: true,
    grantApplicationWithdrawn: true,
    newCommentOnApplication: true,
    applicationTransfer: true,
    applicationStatusChange: true,
    //Project
    projectUserInvite: true,
    projectUserRemoved: true,
    verificationReminder: true,
    userRoleChange: true,
    userLeft: true,
    lowFunds: true
}

export enum MailType {
    NEW_GRANT_APPLICATION,
    GRANT_AUTO_APPROVE,
    GRANT_APPLICATION_UPDATED,
    GRANT_APPLICATION_APPROVED,
    GRANT_APPLICATION_REJECTED,
    GRANT_APPLICATION_WITHDRAWN,
    NEW_COMMENT_ON_APPLICATION,
    APPLICATION_TRANSFER,
    APPLICATION_STATUS_CHANGE,
    //Project
    PROJECT_USER_INVITE,
    PROJECT_USER_REMOVED,
    VERIFICATION_REMINDER,
    USER_ROLE_CHANGE,
    USER_LEFT,
    LOW_FUNDS
}

const initialState: UserDetailsState = {
    settings: defaultEmailSettings
};

type Action<T, B> = {type: T; payload: B};
type UpdatePlaceholdersEmailSettings = Action<"UpdatePlaceholdersEmailSettings", UserDetailsState>;

const reducer = (state: UserDetailsState, action: UpdatePlaceholdersEmailSettings): UserDetailsState => {
    switch (action.type) {
        case "UpdatePlaceholdersEmailSettings":
            return {...state, ...action.payload};
    }
};

export const ChangeEmailSettings: React.FunctionComponent<{setLoading: (loading: boolean) => void}> = () => {
    const [commandLoading, invokeCommand] = useCloudCommand();
    const [state, dispatch] = React.useReducer(reducer, initialState, () => initialState);

    var currentEmailSettings: EmailSettings = defaultEmailSettings
    const info = useCallback(async () => {

        const emailSettings = await invokeCommand(
            retrieveEmailSettings({})
        )

        currentEmailSettings = emailSettings.settings ?? defaultEmailSettings

        dispatch({
            type: "UpdatePlaceholdersEmailSettings",
            payload: {
                settings: currentEmailSettings
            }
        });
    }, []);

    useEffect(() => {
        info();
    }, []);

    const update = useCallback(async () => {
        dispatch({
            type: "UpdatePlaceholdersEmailSettings",
            payload: {
                settings: currentEmailSettings
            }
        })
    }, [])

    const onSubmit = useCallback(async (e: React.SyntheticEvent) => {
        e.preventDefault();

        if (commandLoading) return;

        const wasSuccessful = await invokeCommand(toggleEmailSettings(bulkRequestOf({
            settings: currentEmailSettings
        }))) !== null;

        if (!wasSuccessful) {
            snackbarStore.addFailure("Failed to update user email settings", false);
        } else {
            snackbarStore.addSuccess("User email settings updated", false);
        }
    }, [commandLoading]);

    function toggleSubscription(type: MailType) {
        switch (type) {
            case MailType.NEW_GRANT_APPLICATION:
                state.settings.newGrantApplication = !state.settings.newGrantApplication
                currentEmailSettings.newGrantApplication = state.settings.newGrantApplication
                break;
            case MailType.GRANT_AUTO_APPROVE:
                state.settings.grantAutoApprove = !state.settings.grantAutoApprove
                currentEmailSettings.grantAutoApprove = state.settings.grantAutoApprove
                break;
            case MailType.GRANT_APPLICATION_UPDATED:
                state.settings.grantApplicationUpdated = !state.settings.grantApplicationUpdated
                currentEmailSettings.grantApplicationUpdated = state.settings.grantApplicationUpdated
                break;
            case MailType.GRANT_APPLICATION_APPROVED:
                state.settings.grantApplicationApproved = !state.settings.grantApplicationApproved
                currentEmailSettings.grantApplicationApproved = state.settings.grantApplicationApproved
                break;
            case MailType.GRANT_APPLICATION_REJECTED:
                state.settings.grantApplicationRejected = !state.settings.grantApplicationRejected
                currentEmailSettings.grantApplicationRejected = state.settings.grantApplicationRejected
                break;
            case MailType.GRANT_APPLICATION_WITHDRAWN:
                state.settings.grantApplicationWithdrawn = !state.settings.grantApplicationWithdrawn
                currentEmailSettings.grantApplicationWithdrawn = state.settings.grantApplicationWithdrawn
                break;
            case MailType.NEW_COMMENT_ON_APPLICATION:
                state.settings.newCommentOnApplication = !state.settings.newCommentOnApplication
                currentEmailSettings.newCommentOnApplication = state.settings.newCommentOnApplication
                break;
            case MailType.PROJECT_USER_INVITE:
                state.settings.projectUserInvite = !state.settings.projectUserInvite
                currentEmailSettings.projectUserInvite = state.settings.projectUserInvite
                break;
            case MailType.PROJECT_USER_REMOVED:
                state.settings.projectUserRemoved = !state.settings.projectUserRemoved
                currentEmailSettings.projectUserRemoved = state.settings.projectUserRemoved
                break;
            case MailType.VERIFICATION_REMINDER:
                state.settings.verificationReminder = !state.settings.verificationReminder
                currentEmailSettings.verificationReminder = state.settings.verificationReminder
                break;
            case MailType.USER_ROLE_CHANGE:
                state.settings.userRoleChange = !state.settings.userRoleChange
                currentEmailSettings.userRoleChange = state.settings.userRoleChange
                break;
            case MailType.USER_LEFT:
                state.settings.userLeft = !state.settings.userLeft
                currentEmailSettings.userLeft = state.settings.userLeft
                break;
            case MailType.LOW_FUNDS:
                state.settings.lowFunds = !state.settings.lowFunds
                currentEmailSettings.lowFunds = state.settings.lowFunds
                break;
            case MailType.APPLICATION_STATUS_CHANGE:
                state.settings.applicationStatusChange = !state.settings.applicationStatusChange
                currentEmailSettings.applicationStatusChange = state.settings.applicationStatusChange
                break;
        }
    }

    return (
        <Box mb={16}>
            <Heading.h2>Email Settings</Heading.h2>
            <form onSubmit={onSubmit}>

                <Heading.h5>Grant Applications</Heading.h5>
                <Label ml={10} width="45%" style={{display: "inline-block"}}>
                    <Checkbox
                        size={27}
                        onClick={() => toggleSubscription(MailType.GRANT_APPLICATION_APPROVED)}
                        onChange={update}
                        checked={state.settings.grantApplicationApproved}
                    />
                    <Box as="span">Application Approved</Box>
                </Label>
                <Label ml={10} width="45%" style={{display: "inline-block"}}>
                    <Checkbox
                        size={27}
                        onClick={() => toggleSubscription(MailType.GRANT_APPLICATION_REJECTED)}
                        onChange={update}
                        checked={state.settings.grantApplicationRejected}
                    />
                    <Box as="span">Application Rejected</Box>
                </Label>
                <Label ml={10} width="45%" style={{display: "inline-block"}}>
                    <Checkbox
                        size={27}
                        onClick={() => toggleSubscription(MailType.GRANT_APPLICATION_WITHDRAWN)}
                        onChange={update}
                        checked={state.settings.grantApplicationWithdrawn}
                    />
                    <Box as="span">Application Withdrawn</Box>
                </Label>
                <Label ml={10} width="45%" style={{display: "inline-block"}}>
                    <Checkbox
                        size={27}
                        onClick={() => toggleSubscription(MailType.NEW_GRANT_APPLICATION)}
                        onChange={update}
                        checked={state.settings.newGrantApplication}
                    />
                    <Box as="span">New Application Received</Box>
                </Label>
                <Label ml={10} width="45%" style={{display: "inline-block"}}>
                    <Checkbox
                        size={27}
                        onClick={() => toggleSubscription(MailType.APPLICATION_STATUS_CHANGE)}
                        onChange={update}
                        checked={state.settings.applicationStatusChange}
                    />
                    <Box as="span">Status Change By Others Admins</Box>
                </Label>
                <Label ml={10} width="45%" style={{display: "inline-block"}}>
                    <Checkbox
                        size={27}
                        onClick={() => toggleSubscription(MailType.APPLICATION_TRANSFER)}
                        onChange={update}
                        checked={state.settings.applicationTransfer}
                    />
                    <Box as="span">Transfers From Other Projects</Box>
                </Label>
                <Label ml={10} width="45%" style={{display: "inline-block"}}>
                    <Checkbox
                        size={27}
                        onClick={() => toggleSubscription(MailType.GRANT_AUTO_APPROVE)}
                        onChange={update}
                        checked={state.settings.grantAutoApprove}
                    />
                    <Box as="span">On Auto Approval</Box>
                </Label>
                <Label ml={10} width="45%" style={{display: "inline-block"}}>
                    <Checkbox
                        size={27}
                        onClick={() => toggleSubscription(MailType.NEW_COMMENT_ON_APPLICATION)}
                        onChange={update}
                        checked={state.settings.newCommentOnApplication}
                    />
                    <Box as="span">New Comment In Application</Box>
                </Label>
                <Label ml={10} width="45%" style={{display: "inline-block"}}>
                    <Checkbox
                        size={27}
                        onClick={() => toggleSubscription(MailType.GRANT_APPLICATION_UPDATED)}
                        onChange={update}
                        checked={state.settings.grantApplicationUpdated}
                    />
                    <Box as="span">Application Has Been Edited</Box>
                </Label>

                <Heading.h5>Projects</Heading.h5>
                <Label ml={10} width="45%" style={{display: "inline-block"}}>
                    <Checkbox
                        size={27}
                        onClick={() => toggleSubscription(MailType.LOW_FUNDS)}
                        onChange={update}
                        checked={state.settings.lowFunds}
                    />
                    <Box as="span">Low On Funds</Box>
                </Label>
                <Label ml={10} width="45%" style={{display: "inline-block"}}>
                    <Checkbox
                        size={27}
                        onClick={() => toggleSubscription(MailType.PROJECT_USER_INVITE)}
                        onChange={update}
                        checked={state.settings.projectUserInvite}
                    />
                    <Box as="span">User Invited To Project</Box>
                </Label>
                <Label ml={10} width="45%" style={{display: "inline-block"}}>
                    <Checkbox
                        size={27}
                        onClick={() => toggleSubscription(MailType.USER_LEFT)}
                        onChange={update}
                        checked={state.settings.userLeft}
                    />
                    <Box as="span">User Left Project</Box>
                </Label>
                <Label ml={10} width="45%" style={{display: "inline-block"}}>
                    <Checkbox
                        size={27}
                        onClick={() => toggleSubscription(MailType.USER_ROLE_CHANGE)}
                        onChange={update}
                        checked={state.settings.userRoleChange}
                    />
                    <Box as="span">User Role Changed</Box>
                </Label>
                <Label ml={10} width="45%" style={{display: "inline-block"}}>
                    <Checkbox
                        size={27}
                        onClick={() => toggleSubscription(MailType.PROJECT_USER_REMOVED)}
                        onChange={update}
                        checked={state.settings.projectUserRemoved}
                    />
                    <Box as="span">User Removed From Project</Box>
                </Label>
                <Label ml={10} width="45%" style={{display: "inline-block"}}>
                    <Checkbox
                        size={27}
                        onClick={() => toggleSubscription(MailType.VERIFICATION_REMINDER)}
                        onChange={update}
                        checked={state.settings.verificationReminder}
                    />
                    <Box as="span">Verification Reminders</Box>
                </Label>

                <Heading.h5> </Heading.h5>
                <Button
                    mt="1em"
                    type="submit"
                    color="green"
                    disabled={commandLoading}
                >
                    Update Email Settings
                </Button>
            </form>
        </Box>
    );
};
