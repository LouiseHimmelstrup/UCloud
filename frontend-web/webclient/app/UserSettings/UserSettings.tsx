import {Client} from "@/Authentication/HttpClientInstance";
import {MainContainer} from "@/ui-components/MainContainer";
import {setLoading, useTitle} from "@/Navigation/Redux";
import * as React from "react";
import {useDispatch, useSelector} from "react-redux";
import {Box, Flex} from "@/ui-components";
import * as Heading from "@/ui-components/Heading";
import {ChangePassword} from "@/UserSettings/ChangePassword";
import {Sessions} from "@/UserSettings/Sessions";
import {TwoFactorSetup} from "./TwoFactorSetup";
import {ChangeOptionalUserDetails, ChangeUserDetails} from "@/UserSettings/ChangeUserDetails";
import {ChangeEmailSettings} from "@/UserSettings/ChangeEmailSettings";
import {CustomTheming} from "./CustomTheme";
import {refreshFunctionCache} from "@/Utilities/ReduxUtilities";

function UserSettings(): React.ReactNode {

    useTitle("User Settings");

    const headerLoading = useSelector(({status}: ReduxObject) => status.loading);
    const dispatch = useDispatch();

    const setHeaderLoading = React.useCallback((loading: boolean) => {
        dispatch(setLoading(loading));
    }, [dispatch]);

    const mustActivate2fa =
        Client.userInfo?.twoFactorAuthentication === false &&
        Client.userInfo?.principalType === "password";

    return (
        <Flex alignItems="center" flexDirection="column">
            <Box width={0.7}>
                <MainContainer
                    header={<Heading.h1>User Settings</Heading.h1>}
                    main={(
                        <>
                            <TwoFactorSetup
                                mustActivate2fa={mustActivate2fa}
                                loading={headerLoading}
                                setLoading={setHeaderLoading}
                            />

                            {mustActivate2fa ? null : (
                                <>
                                    <ChangePassword
                                        setLoading={setHeaderLoading}
                                    />

                                    <ChangeUserDetails
                                        setLoading={setHeaderLoading}
                                    />
                                    <ChangeOptionalUserDetails />
                                    <ChangeEmailSettings
                                        setLoading={setHeaderLoading}
                                    />
                                    <Sessions
                                        setLoading={setHeaderLoading}
                                        setRefresh={fn => refreshFunctionCache.setRefreshFunction(fn ?? (() => undefined))}
                                    />
                                    <CustomTheming />
                                </>
                            )}

                        </>
                    )}
                />
            </Box>
        </Flex>
    );
}

export default UserSettings;
