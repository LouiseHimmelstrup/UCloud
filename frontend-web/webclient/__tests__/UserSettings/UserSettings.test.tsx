import * as React from "react";
import UserSettings from "UserSettings/UserSettings";
import { create } from "react-test-renderer";
import { mount, configure } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import { FormField, Button } from "semantic-ui-react";

configure({ adapter: new Adapter() });

describe("UserSettings", () => {
    test("Mount", () => {
        expect(create(<UserSettings />).toJSON()).toMatchSnapshot()
    });

    test("Update current password field", () => {
        const userSettings = mount(<UserSettings />);
        userSettings.find(FormField).findWhere(it => it.props().placeholder === "Old password").find("input").simulate("change", { target: { value: "current password" } });
        expect(userSettings.state("currentPassword")).toBe("current password")
    });

    test("Update new password field", () => {
        const userSettings = mount(<UserSettings />);
        userSettings.find(FormField).findWhere(it => it.props().placeholder === "New password").find("input").simulate("change", { target: { value: "new password" } });
        expect(userSettings.state("newPassword")).toBe("new password")
    });

    test("Update repeated password field", () => {
        const userSettings = mount(<UserSettings />);
        userSettings.find(FormField).findWhere(it => it.props().placeholder === "Repeat password").find("input").simulate("change", { target: { value: "repeated password" } });
        expect(userSettings.state("repeatedPassword")).toBe("repeated password")
    });

    test("Submit with missing old password", () => {
        const userSettings = mount(<UserSettings />);
        const password = "password";
        userSettings.find(FormField).findWhere(it => it.props().placeholder === "New password").find("input").simulate("change", { target: { value: password } });
        userSettings.find(FormField).findWhere(it => it.props().placeholder === "Repeat password").find("input").simulate("change", { target: { value: password } });
        userSettings.find(Button).findWhere(it => it.props().content === "Change password").simulate("submit");
        expect(userSettings.state("error")).toBe(true);
    });

    test("Submit with missing password", () => {
        const userSettings = mount(<UserSettings />);
        const password = "password";
        userSettings.find(FormField).findWhere(it => it.props().placeholder === "Old password").find("input").simulate("change", { target: { value: "current password" } });
        userSettings.find(FormField).findWhere(it => it.props().placeholder === "Repeat password").find("input").simulate("change", { target: { value: password } });
        userSettings.find(Button).findWhere(it => it.props().content === "Change password").simulate("submit");
        expect(userSettings.state("error")).toBe(true);
    });

    test("Submit with missing repeated password", () => {
        const userSettings = mount(<UserSettings />);
        const password = "password";
        userSettings.find(FormField).findWhere(it => it.props().placeholder === "Old password").find("input").simulate("change", { target: { value: "current password" } });
        userSettings.find(FormField).findWhere(it => it.props().placeholder === "New password").find("input").simulate("change", { target: { value: password } });
        userSettings.find(Button).findWhere(it => it.props().content === "Change password").simulate("submit");
        expect(userSettings.state("error")).toBe(true);
    });

    test("Submit with non-matching new passwords", () => {
        const userSettings = mount(<UserSettings />);
        const password = "password";
        userSettings.find(FormField).findWhere(it => it.props().placeholder === "Old password").find("input").simulate("change", { target: { value: "current password" } });
        userSettings.find(FormField).findWhere(it => it.props().placeholder === "New password").find("input").simulate("change", { target: { value: "one password" } });
        userSettings.find(FormField).findWhere(it => it.props().placeholder === "Repeat password").find("input").simulate("change", { target: { value: "another password" } });
        userSettings.find(Button).findWhere(it => it.props().content === "Change password").simulate("submit");
        expect(userSettings.state("error") && userSettings.state("repeatPasswordError")).toBe(true);
    });
});