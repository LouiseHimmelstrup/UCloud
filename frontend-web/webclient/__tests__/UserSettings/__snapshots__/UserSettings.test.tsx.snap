// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserSettings Mount 1`] = `
<div
  className="sc-bZQynM Ygzij"
>
  <div
    className="sc-htpNat hwVXPq"
    cursor="auto"
    width={0.7}
  >
    <div
      className="sc-htpNat bWSHjv"
      cursor="auto"
    >
      <h1
        className="sc-bdVaJa jOPwGr"
        cursor="inherit"
        fontSize={6}
      >
        Change Password
      </h1>
    </div>
    <div
      className="sc-bZQynM bTpYSl"
    >
      <div
        className="sc-htpNat dCbdhM"
        cursor="auto"
        width={1}
      >
        <div
          className="sc-htpNat sc-jTzLTM kNEtzm"
        />
        <form
          onSubmit={[Function]}
        >
          <div
            className="sc-htpNat dQLcuo"
            cursor="auto"
          >
            <label
              className="sc-ckVGcZ gXhFce"
              color="black"
              fontSize={1}
              fontWeight="bold"
            >
              Current Password
              <input
                className="sc-chPdSV ktXeRQ"
                id="default"
                onChange={[Function]}
                placeholder="Current password"
                type="password"
                value=""
              />
            </label>
          </div>
          <div
            className="sc-htpNat dQLcuo"
            cursor="auto"
          >
            <label
              className="sc-ckVGcZ gXhFce"
              color="black"
              fontSize={1}
              fontWeight="bold"
            >
              New Password
              <input
                className="sc-chPdSV ktXeRQ"
                id="default"
                onChange={[Function]}
                placeholder="New password"
                type="password"
                value=""
              />
            </label>
          </div>
          <div
            className="sc-htpNat dQLcuo"
            cursor="auto"
          >
            <label
              className="sc-ckVGcZ gXhFce"
              color="black"
              fontSize={1}
              fontWeight="bold"
            >
              Repeat new password
              <input
                className="sc-chPdSV ktXeRQ"
                id="default"
                onChange={[Function]}
                placeholder="Repeat password"
                type="password"
                value=""
              />
            </label>
          </div>
          <button
            className="sc-htoDjs hpPnDS"
            color="green"
            type="submit"
          >
            Change password
          </button>
        </form>
        <h1
          className="sc-bdVaJa jOPwGr"
          cursor="inherit"
          fontSize={6}
        >
          Two Factor Authentication
        </h1>
        <b>
          You do NOT have a 2FA device connected to your account
        </b>
        <hr
          className="sc-VigVT kXJjWb"
        />
        <div>
          <h3
            className="sc-bdVaJa bNfRTK"
            cursor="inherit"
            fontSize={4}
          >
            2FA Setup
          </h3>
          <p>
            In order to activate 2FA on your account you must have a device capable of issuing time-based one time passwords (TOTP). For this we recommend the "Google Authenticator" app, available for both Android and iOS.
          </p>
          <div
            className="sc-bZQynM bqIUrb"
          >
            <a
              href="https://play.google.com/store/apps/details?id=com.google.android.apps.authenticator2&hl=en_us"
              rel="noopener"
              target="_blank"
            >
              <img
                className="sc-kAzzGY kHvfeN"
                src={[Function]}
                width="150px"
              />
            </a>
            <a
              href="https://itunes.apple.com/us/app/google-authenticator/id388497605"
              rel="noopener"
              target="_blank"
            >
              <img
                className="sc-kAzzGY kHvfeN"
                src={[Function]}
                width="150px"
              />
            </a>
          </div>
          <p>
            Once you are ready click the button below to get started:
          </p>
          <button
            className="sc-htoDjs dacUeo"
            color="green"
            content="Start setup"
            disabled={true}
            onClick={[Function]}
          >
            <svg
              className="sc-gzVnrw kudwwP"
              cursor="inherit"
              fill="currentcolor"
              height={24}
              id="outer_ellipsis_svg__Layer_1"
              viewBox="0 0 24 24"
              width={24}
              x={0}
              xmlSpace="preserve"
              y={0}
            >
              <style />
              <ellipse
                cx={5}
                cy={12}
                rx={3}
                ry={3}
              />
              <ellipse
                cx={19}
                cy={12}
                rx={3}
                ry={3}
              />
            </svg>
          </button>
        </div>
      </div>
      <div
        className="sc-htpNat sc-jTzLTM eGihWZ"
        width={
          Array [
            0,
            0,
          ]
        }
      >
        <div
          className="sc-htpNat fsPsYE"
          cursor="auto"
        />
      </div>
    </div>
  </div>
</div>
`;
